yew/src/sgml_tags.rs

```diff
@@ -0,0 +1,230 @@
+fn is_control(c: char) -> bool {
+    match c {
+        '\u{007F}'..='\u{009F}' => true,
+        _ => is_c0_control(c),
+    }
+}
+
+/// Returns true when the character provided is a "c0 control" as defined
+/// in [the WhatWG spec](https://infra.spec.whatwg.org/#c0-control)
+fn is_c0_control(c: char) -> bool {
+    match c {
+        '\u{0000}'..='\u{001F}' => true,
+        _ => false,
+    }
+}
+
+/// Returns true when the string provided is a "noncharacter" as defined
+/// in [the WhatWG spec](https://infra.spec.whatwg.org/#noncharacter)
+fn is_noncharacter(c: char) -> bool {
+    match c {
+        '\u{FDD0}'..='\u{FDEF}' => true,
+        '\u{FFFE}' | '\u{FFFF}' | '\u{1FFFE}' | '\u{1FFFF}' | '\u{2FFFE}' | '\u{2FFFF}'
+        | '\u{3FFFE}' | '\u{3FFFF}' | '\u{4FFFE}' | '\u{4FFFF}' | '\u{5FFFE}' | '\u{5FFFF}'
+        | '\u{6FFFE}' | '\u{6FFFF}' | '\u{7FFFE}' | '\u{7FFFF}' | '\u{8FFFE}' | '\u{8FFFF}'
+        | '\u{9FFFE}' | '\u{9FFFF}' | '\u{AFFFE}' | '\u{AFFFF}' | '\u{BFFFE}' | '\u{BFFFF}'
+        | '\u{CFFFE}' | '\u{CFFFF}' | '\u{DFFFE}' | '\u{DFFFF}' | '\u{EFFFE}' | '\u{EFFFF}'
+        | '\u{FFFFE}' | '\u{FFFFF}' | '\u{10FFFE}' | '\u{10FFFF}' => true,
+        _ => false,
+    }
+}
+
+/// Returns true when the string provided is a valid "attribute name" as defined
+/// in [the WhatWG spec](https://html.spec.whatwg.org/multipage/syntax.html#syntax-attribute-name)
+pub fn is_valid_html_attribute_name(attr: &str) -> bool {
+    for c in attr.chars() {
+        if is_noncharacter(c)
+            || is_control(c)
+            || c == '\u{0020}'
+            || c == '\u{0022}'
+            || c == '\u{0027}'
+            || c == '\u{003E}'
+            || c == '\u{002F}'
+            || c == '\u{003D}'
+        {
+            return false;
+        }
+    }
+    true
+}
+
+/// Returns true when the character provided is a valid PCENChar as defined
+/// in [the WhatWG spec](https://html.spec.whatwg.org/multipage/custom-elements.html#prod-pcenchar)
+fn is_pcen_char(c: char) -> bool {
+    match c {
+        '-' | '.' | '0'..='9' | 'a'..='z' | '_' => true,
+        '\u{B7}' => true,
+        '\u{C0}'..='\u{D6}' => true,
+        '\u{D8}'..='\u{F6}' => true,
+        '\u{F8}'..='\u{37D}' => true,
+        '\u{37F}'..='\u{1FFF}' => true,
+        '\u{200C}'..='\u{200D}' => true,
+        '\u{203F}'..='\u{2040}' => true,
+        '\u{2070}'..='\u{218F}' => true,
+        '\u{2C00}'..='\u{2FEF}' => true,
+        '\u{3001}'..='\u{D7FF}' => true,
+        '\u{F900}'..='\u{FDCF}' => true,
+        '\u{FDF0}'..='\u{FFFD}' => true,
+        '\u{10000}'..='\u{EFFFF}' => true,
+        _ => false,
+    }
+}
+
+/// Returns true when the tag name provided would be a valid "custom element" per
+/// [the WhatWG spec](https://html.spec.whatwg.org/multipage/custom-elements.html#valid-custom-element-name)
+fn is_valid_html_custom_element_name(tag: &str) -> bool {
+    if DISALLOWED_CUSTOM_ELEMENT_TAGS.contains(&tag) {
```

Comment:
I don't think we need this check. Yes, these names must not be used for custom elements but that's only because they are already reserved for non-custom tags.

This function isn't public and is only exposed through `is_valid_sgml_tag` which already accepts all the tag names in the list in the `is_valid_html_standard_ish_element_name` check. So currently this check will never return false.

We can add a note to the documentation to make it clear that this function doesn't handle disallowed custom element tags.

Reply:
Since the public interface is agnostic to these tags, I see no reason to remove the code. The function is meant to represent the way WhatWG defines a custom tag; to stray outside of that bounds would set a precedent in the codebase for not matching the spec exactly, which is a fraught path to go down when writing code that's meant to be secure.

Reply:
While I would usually agree, in this case it makes absolutely no sense to incur the performance and space penalty. This is just dead code that Rust can't eliminate.
Currently it's also entirely irrelevant to Yew whether it's a custom element tag or a deprecated tag containing a hyphen.
I totally understand your stance on adhering to the spec - after all, that's the reason we went down this rabbit hole in the first place. But sometimes being pedantic doesn't yield any benefits.
Adding a note to the documentation should also sufficiently take care of your concerns about precedent. Maybe we can even rename the function to `contains_valid_html_custom_element_characters` to ease your mind?